vault/expiration.go

```diff
@@ -2291,6 +2316,47 @@ func (m *ExpirationManager) walkLeases(walkFn leaseWalkFunction) error {
 	return nil
 }
 
+// must be called with m.pendingLock held
+func (m *ExpirationManager) removeFromPending(ctx context.Context, leaseID string) {
+	if info, ok := m.pending.Load(leaseID); ok {
+		pending := info.(pendingInfo)
+		pending.timer.Stop()
+		m.pending.Delete(leaseID)
+		m.leaseCount--
+		// Log but do not fail; unit tests (and maybe Tidy on production systems)
+		if err := m.core.quotasHandleLeases(ctx, quotas.LeaseActionDeleted, []string{leaseID}); err != nil {
+			m.logger.Error("failed to update quota on revocation", "error", err)
+		}
+	}
+}
+
+// note: must be called with pending lock held
+func (m *ExpirationManager) markLeaseAsZombie(ctx context.Context, le *leaseEntry, err error) {
+	if le == nil {
+		m.logger.Warn("attempted to mark nil lease as zombie")
+		return
+	}
+	if le.isZombie() {
+		m.logger.Info("attempted to re-mark lease as zombie", "original_error", le.RevokeErr, "new_error", err.Error())
+		return
+	}
+
+	errStr := err.Error()
+	if len(errStr) == 0 {
+		errStr = "no error message given"
+	}
+	if len(errStr) > maxZombieErrorLength {
+		errStr = errStr[:maxZombieErrorLength]
+	}
+
+	le.RevokeErr = errStr
+	m.persistEntry(ctx, le)
+
+	m.zombies.Store(le.LeaseID, le)
+	m.removeFromPending(ctx, le.LeaseID)
```

Comment:
If we remove from pending here we are no longer counting these leases in the `num_leases` metric and they no longer count towards lease quotas. Do you have thoughts on how zombie leases should interact with those two systems?

Reply:
yes. i'm now thinking that zombies should be included in the lease count quota (mostly for storage reasons). i need to do some discovery on the `(ExpirationManager).leaseCount` system